Khang.truong/on missing data guide (#26115)
DataDog/documentation: content/en/monitors/guide/on_missing_data.md

---
title: Migrating to On Missing Data Configuration
further_reading:
- link: "/api/latest/monitors/"
  tag: "API"
  text: "Monitors API Documentation"
---

## Overview

> Metric Monitors offer enhanced options for handling missing data, allowing you to differentiate between missing data as a failure mode and a healthy state. 

Reviewer: ```suggestion
Metric monitors offer enhanced options for handling missing data, allowing you to differentiate between missing data as a failure mode and a healthy state. 
```

These options align with what's available in other monitors types like Logs, Events, CI, Database, Error Tracking, and more.

## Benefits of using On Missing Data options

> When measuring the number of bad events, like errors, Monitors should reflect an "OK" when no data is detected. With the legacy No Data configurations, monitors would report No Data. The On Missing Data configuration options allows monitors to reflect health states more accurately, improving clarity.

Reviewer: I think making it all report instead of report vs reflect makes the comparison clearer.

```suggestion
The legacy No Data configurations cause monitors to report No Data. When measuring the number of bad events, like errors, monitors should report an "OK" status when no data is detected. The On Missing Data configuration options allow monitors to report health states more accurately, improving clarity.
```

Author: I feel like having a side by side comparaison of Expected behaviour VS Legacy behaviour makes it more digestable by the user

## Monitors managed through the UI

> If you manage your monitors from the UI, they will be automatically updated for you the next time you edit them. If you'd like to update them sooner, you can do it through the API as explained below.

Reviewer: ```suggestion
If you manage your monitors from the UI, the configuration automatically updates the next time you edit them. To update the On Missing Data configuration sooner, see the following sections on adjusting through the API.
```

---
title: Migrating to On Missing Data Configuration
further_reading:
- link: "/api/latest/monitors/"
  tag: "API"
  text: "Monitors API Documentation"
---

## Overview

Metric Monitors offer enhanced options for handling missing data, allowing you to differentiate between missing data as a failure mode and a healthy state. 

> These options align with what's available in other monitors types like CI, Database, Error Tracking, and more. They apply to query-based monitors, including Integration and Anomaly monitors, providing two main configurations based on the aggregation type:

Reviewer 2: ```suggestion
These options align with what's available in other monitors types like Logs, Events, CI, Database, Error Tracking, and more.
```

## Benefits of using On Missing Data options

Metric Monitors tracking errors should reflect as "OK" when no errors occur. With the legacy No Data configurations, monitors would raise alerts with missing data, leading to confusion and workarounds. The On Missing Data configuration options allow monitors to reflect health states, eliminating the need for workarounds.

## Monitors managed through the UI

> If you manage your monitors from the UI, they are automatically updated for you. If you'd like to update them sooner, you can do it through the API as explained below.

Reviewer 2: ```suggestion
If you manage your monitors from the UI, they will be  automatically updated for you the next time you edit them. If you'd like to update them sooner, you can do it through the API as explained below.
```

> If you manage your monitors from the UI, they will be automatically updated for you. If you'd like to update them sooner, you can do it through the API as explained below.

Reviewer: ```suggestion
If you manage your monitors from the UI, they are automatically updated for you. If you'd like to update them sooner, you can do it through the API as explained below.
```

## Monitors managed through the API or Terraform

> If you are managing your monitors with API or Terraform, replace `notify_no_data` and `no_data_timeframe` with `on_missing_data`. The `no_data_timeframe` parameter is not required since `on_missing_data` uses the same timeframe as the time window.  

Reviewer: ```suggestion
If you are managing your monitors with API or Terraform, replace `notify_no_data` and `no_data_timeframe` with `on_missing_data`. The `no_data_timeframe` parameter is not required since `on_missing_data` uses the same time frame as the time window.  
```

The available values for `on_missing_data` are:

* default  
* show_no_data  
* show_and_notify_no_data  

> * resolve

Reviewer 2: This feel redundant with the next section. wdyt?

Author: Fair point

> ### API Parameters

Reviewer: ```suggestion
### API parameters
```

The previous No Data parameter, `notify_no_data`, remains available on existing monitors and are not automatically upgraded to the new `on_missing_data` features.

| Parameter                               | UI Description                                                                                     |
|-----------------------------------------|----------------------------------------------------------------------------------------------------|
| `"on_missing_data": "show_and_notify_no_data"` | If data is missing Show NO DATA and notify<br>(Formerly, "Notify if data is missing")                       |
| `"on_missing_data": "show_no_data"`     | If data is missing Show NO DATA<br>(Formerly, "Do not notify if data is missing")                           |
| `"on_missing_data": "resolve"`          | If data is missing Show OK                                                                       |
| `"on_missing_data": "default"` if using sum or count aggregation | If data is missing Evaluate as 0 (or other default value)                                  |

> | `"on_missing_data": "default"` if using all other aggregation types | If data is missing Show last known status (See the table above to determine which will be enabled) |

Reviewer: ```suggestion
| `"on_missing_data": "default"` if using all other aggregation types | If data is missing Show last known status |
```

For all the available fields, see the [API Documentation][1].

Here's an example of before and after of a JSON monitor with those fields:

**Before**  
```json
{  
  "name": "CPU usage is high for host $host.value",  
    "type": "query alert",  
    "query": "avg(last_5m):100 - avg:system.cpu.idle{$host} > 90",  
    "message": "A high CPU usage has been detected for host $host.value, which can impact the system performance.",  
    "tags": [],  
    "options": {  
        "thresholds": { "critical": 90 },  
        "notify_audit": false,  
        "include_tags": false,  
        "notify_no_data": true,  

>         "no_data_timeframe": 10  

Reviewer 2: Let's highlight these rows as well.

  "name": "CPU usage is high for host $host.value",  
    "type": "query alert",  
    "query": "avg(last_5m):100 - avg:system.cpu.idle{$host} > 90",  
    "message": "A high CPU usage has been detected for host $host.value, which can impact the system performance.",  
    "tags": [],  
    "options": {  
       "thresholds": { "critical": 90 },  
       "notify_audit": false,  
       "include_tags": false,  
       "on_missing_data": "show_and_notify_no_data"  
    }  
}  
{{< /highlight >}}

## SLO monitors

SLOs treat uptime and downtime according to this mapping:

| On Missing Data Configuration | Monitor Status                 | SLO Treatment               |
|-------------------------------|--------------------------------|-----------------------------|
| Show OK                       | OK                             | Uptime                      |
| Show No Data                  | No Data                        | Uptime                      |
| Show No Data and Notify       | No Data                        | Downtime                    |
| Show last known status        | Whatever the last status was   | If OK, Uptime<br>If Alert, Downtime |

> | Evaluate as zero              | Depends on the threshold configuration | If OK, Uptime<br>If Alert, Downtime |

Reviewer 2: @roxanne-moslehi any additional thoughts to add?

Author: Resolving for now
